### PR TITLE
tags: initial implemention of post tagging

### PR DESCRIPTION
**Pull request by taskboy3000**
This patch adds the facility to plerd for tagging blog posts and
creating tag indexes of same.

* new attribute 'tags' supported in blog post source files
* tags are a comma-seperated list (quoting is not required or
  supported)
* tag index added to nav bar in wrapper.tt
* new tag index view shows all tags and their associated posts
* tag NAME specific views are created for each tag
* within templates, plerd.tag_uri(TAGNAME) can be use to get URI for a
  given tag
* updated unit test to support tags

### CODE REVIEW

File: lib/Plerd/Post.pm [hunk 1]
```diff
@@ -512,7 +525,9 @@ sub _process_source_file {
     if ( $attributes_need_to_be_written_out ) {
         my $new_content = '';
         for my $attribute_name ( @ordered_attribute_names ) {
-            $new_content .= "$attribute_name: $attributes{ $attribute_name }\n";
```
Comment (jmacdotorg) [inline lib/Plerd/Post.pm:515]: This change seems unrelated to the tag-specific work?

File: t/basic.t [hunk 2]
```diff
@@ -54,8 +54,9 @@ unlink "$FindBin::Bin/source/no-title.md";
 
 $plerd->publish_all;
 
-# The "+3" below accounts for the generated recent, archive, and RSS files.
-my $expected_docroot_count = scalar( $source_dir->children( no_hidden => 1 ) ) + 3;
+# The "+3" below accounts for the generated recent, archive, tag
```
Comment (jmacdotorg) [inline t/basic.t]: It's +4, now. :)

File: t/templates/_tag.tt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<section>
```
Comment (jmacdotorg) [inline t/templates/_tag.tt]: I would rather that this file just be a [% BLOCK %] section in tags.tt, the only template that invokes it.

Using a sub-template file like this isn't bad at all, but making it into a TT-block would be more consistent with the existing example templates.

File: templates/post.tt [hunk 4]
```diff
@@ -125,8 +135,3 @@
         (None yet!)
     [% END %]
 [% END %]
-
-
-<style>
```
Comment (jmacdotorg) [inline templates/post.tt:130]: Why remove this?

File: t/basic.t (hunk 2 above)
Comment (jmacdotorg) [reply t/basic.t]: (Also, the list of stuff makes it sound like it should be +6?)

File: t/templates/wrapper.tt [hunk 5]
```diff
@@ -3,11 +3,23 @@
 <head>
     <title>[% title %]</title>
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
-    <link href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">
-    <link href="atom.xml" rel="alternate" title="[% plerd.title %]" type="application/atom+xml">
+    <link href="https://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">
```
Comment (jmacdotorg) [inline t/templates/wrapper.tt]: Why add all this to the t/ templates now?

(It may be a good idea, but it's unrelated to the tag work...)

File: t/templates/post.tt [hunk 6]
```diff
@@ -110,7 +110,3 @@
     [% END %]
 [% END %]
 
-
```
Comment (jmacdotorg) [inline t/templates/post.tt]: Removing this is probably a good idea, but it's not related to the tag work.

File: t/templates/wrapper.tt (hunk 5 above)
Comment (jmacdotorg) [reply t/templates/wrapper.tt]: Oh, it looks like you pasted in wrapper.tt from templates/wrapper.tt.

I don't think that's at all necessary just to make the tests work, right? Go ahead and just revert this file, please.

File: templates/tags.tt [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+</section>
+[%   ELSE %]
+
+    [% FOREACH tag = tags.keys.sort %]
+        [% INCLUDE _tag.tt tag=tag posts => tags.$tag plerd=plerd %]
+    [% END %]
+
+[%   END %]
+
+[% END %]
```
Comment (jmacdotorg) [inline templates/tags.tt:26]: See comments on _tag.tt about adding  [% BLOCK %] to this file.

File: templates/_tag.tt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<section>
+    <h3><a href="[% plerd.tag_uri(tag) %]">[% tag %]</a></h3>
```
Comment (jmacdotorg) [inline templates/_tag.tt]: As written, this causes a tag-specific page to link to itself via its sole header, which is uneccesary.

File: templates/tags.tt [hunk 9]
```diff
@@ -0,0 +1,20 @@
+[% WRAPPER wrapper.tt title = 'Tags' %]
```
Comment (jmacdotorg) [inline templates/tags.tt:1]: The tag-specific pages look good, but I'm not sure about the top-level tag index. Will post thoughts on the overall review.

File: lib/Plerd.pm [hunk 10]
```diff
@@ -569,6 +650,70 @@ sub generates_post_guids {
...
+
+    return Path::Class::File->new(
+        $self->template_directory,
+        'tags.tt',
+    );
+}
+
+# Return either the tags/index.html file
+# or a tags/TAGNAME.html file if given a tag
+sub tags_publication_file {
```
Comment (jmacdotorg) [inline lib/Plerd.pm:678]: All these public methods ought to be documented in the appropriate section of the POD at the end of this file.

File: lib/Plerd.pm [hunk 11]
```diff
@@ -40,13 +40,26 @@ has 'database_path' => (
     isa => 'Str',
 );
 
+has 'tags_publication_path' => (
```
Comment (jmacdotorg) [inline lib/Plerd.pm:43]: All these public attributes ought to be documented in the appropriate section of the POD at the end of this file.

File: lib/Plerd/Post.pm [hunk 12]
```diff
@@ -60,6 +60,12 @@ has 'attributes' => (
     isa => 'HashRef',
 );
 
+has 'tags' => (
```
Comment (jmacdotorg) [inline lib/Plerd/Post.pm:63]: This attribute ought to be documented in the appropriate section of the POD at the end of this file.

**Review by jmacdotorg — CHANGES_REQUESTED** [review]
This is great! I'm requesting a pile of changes but they're mostly little stuff.

One heavier bit: I think I'd prefer the top-level tag index page to be list a simple list of tags, maybe with a parenthesized count of the matching posts next to each, rather than a fully expanded list of every tag's posts. That would open the door to many (even most) posts getting listed again and again on that page, and I don't like that thought. A nice clean list seems better.

(The tag-specific pages seem fine though!)

**Comment by jmacdotorg** [comment]
Two more comments:

1. The tags' HTML should use `p-category` Microformats2 markup, per <http://microformats.org/wiki/h-entry>.

1. Links to pages with punctuation or emoji don't work, though the tag-index pages for them seem to be created okay, with URI-encoded names. For example: `😀` or `"Foo bar"`. (There should be at least one test for these, as well.)

File: lib/Plerd/Post.pm (hunk 1 above)
Comment (taskboy3000) [reply lib/Plerd/Post.pm:515]: This change was needed to silence warnings in the test.  This wasn't a delete, but an existence/defined check (see below).   There is no harm input validation, is there?

**Comment by taskboy3000** [comment]
1. I thought the blog posts were markup correctly using the h-entry stuff.  Didn't you point me to a validator?  However, I think I should just revert the wrapper.  I admit that I blanket copied the templates.

**Comment by jmacdotorg** [comment]
Yes, the blog posts validate as `h-entry`s. I suggest that we take the opportunity of adding tags to further enrich the existing Microformats metadata, since `h-entry` has a property called `p-category` tailor-made for this.

So, we could e.g. change post.tt so that tags are formatted this way (adding only one new class to the code you've already written):

`<li class="list-inline-item p-category"><a href="[% plerd.tag_uri(tag) %]">[% tag %]</a></li>`

Make sense?

**Comment by taskboy3000** [comment]
Yes, that makes sense.  Will tag the, er, tags.

--Joe

File: lib/Plerd/Post.pm (hunk 1 above)
Comment (jmacdotorg) [reply lib/Plerd/Post.pm:515]: Okay, makes sense, thanks.

May as well just collapse the if-check to `defined $attributes{ $attribute_name }` though, right?

**Comment by taskboy3000** [comment]
I thought that in some old Perl, that defined check would autovivicate the key.  I can’t reproduce that now, so it seems harmless to reduce the check to simply defined.

--Joe

File: lib/Plerd.pm [hunk 13]
```diff
@@ -913,6 +918,21 @@ the blog's docroot -- in other words, the place Plerd will write HTML and XML fi
...
+
+A L<Path::Class::Dir> object representation of the directory within the docroot that holds
+tag index HTML files.
+
+=item tag_index_uri
+
+This is a L<URI> object that points to the tag index.  It is
+particularly helpful when creating navigation.
+
+=item tag_uri TAG
```
Comment (jmacdotorg) [inline lib/Plerd.pm:931]: Please move this to the "OBJECT METHODS" list instead, and format the argument list per the other argument-taking methods found there.

**Review by jmacdotorg — CHANGES_REQUESTED** [review]

**Comment by jmacdotorg** [comment]
Holding off further work on this until #28 (and related tasks) are resolved.

**Comment by jmacdotorg** [comment]
On @taskboy3000's request, I'm taking over further development of this branch. (I plan to request his approval on significant changes I make prior to any merge into the master branch.)

**Comment by jmacdotorg** [comment]
Moving this work into local branch https://github.com/jmacdotorg/plerd/tree/tags. Will open a new pull request based off that after some further field-testing.
